opal/datatype/opal_copy_functions_heterogeneous.c

```diff
@@ -133,6 +164,426 @@ static inline void opal_dt_swap_long_double(void *to_p, const void *from_p, cons
+{
+    unsigned s,e,f[4],f0,f1;
+    union fp_float80 ul;
+    union fp_float128 uq;
+    int f80_is_aligned;
+
+    f80_is_aligned = 1;
+    if ((uintptr_t)f80_buf_to & 0xF) {
+        f80_is_aligned = 0;
+    }
+
+    do {
+        /* input */
+        memcpy(uq.bytes,f128_buf_from,sizeof(uq));
+
+        /* unpack */
+        s = uq.bits.sign;
+        e = uq.bits.exp;
+        f[0] = uq.bits.frac0;
+        f[1] = uq.bits.frac1;
+        f[2] = uq.bits.frac2;
+        f[3] = uq.bits.frac3;
+
+        /* truncate */
+        f1 = (f[3] << 15) | (f[2] >> 17);
+        f0 = (f[2] << 15) | (f[1] >> 17);
+
+        /* implicit bit */
+        if (e)
+            f1 |= (1 << 31);
+        else
+            f1 &= ~(1 << 31);
+
+        /* pack */
+        ul.bits.sign  = s;
+        ul.bits.exp   = e;
+        ul.bits.frac0 = f0;
+        ul.bits.frac1 = f1;
+
+        /* output */
+        /* this started as *f80_buf_to = ul.value;
+           but I'm reluctant to assume alignment */
+        if (f80_is_aligned) {
+            *(long double*)f80_buf_to = ul.value;
+        } else {
+            memcpy(f80_buf_to, &ul.value, sizeof(ul));
+        }
+
+        f80_buf_to += to_extent;
+        f128_buf_from += sizeof(uq);
+        count--;
+    } while (count > 0);
+}
+
+#define LDBL_IS_F64(arch) \
+  (                                                                        \
+    (((arch) & OPAL_ARCH_LDMANTDIGISxx) == OPAL_ARCH_LDMANTDIGIS53)        \
+    &&                                                                     \
+    (((arch) & OPAL_ARCH_LDEXPSIZEISxx) == OPAL_ARCH_LDEXPSIZEIS11)        \
+  )
+#define LDBL_IS_F80(arch) \
+  (                                                                        \
+    (((arch) & OPAL_ARCH_LDMANTDIGISxx) == OPAL_ARCH_LDMANTDIGIS64)        \
+    &&                                                                     \
+    (((arch) & OPAL_ARCH_LDEXPSIZEISxx) == OPAL_ARCH_LDEXPSIZEIS15)        \
+  )
+#define LDBL_IS_F128(arch) \
+  (                                                                        \
+    (((arch) & OPAL_ARCH_LDMANTDIGISxx) == OPAL_ARCH_LDMANTDIGIS113)       \
+    &&                                                                     \
+    (((arch) & OPAL_ARCH_LDEXPSIZEISxx) == OPAL_ARCH_LDEXPSIZEIS15)        \
+  )
+#define LDBL_INFO_MASK (OPAL_ARCH_LDMANTDIGISxx | OPAL_ARCH_LDEXPSIZEISxx)
+
+// ldbl_to_f128 (copies a long double(from_arch format) to a float128(local_endian))
+static inline
+void
+ldbl_to_f128(unsigned char *f128_buf_to, const unsigned char *ldbl_buf_from, ssize_t count, int from_arch, ptrdiff_t from_extent)
+{
+    int ldbl_is_aligned;
+
+    ldbl_is_aligned = 1;
+    int alignment_mask = sizeof(long double) - 1;
+    if ((uintptr_t)ldbl_buf_from & alignment_mask) {
+        ldbl_is_aligned = 0;
+    }
+
+#ifdef HAVE___FLOAT128
+    int f128_is_aligned;
```

Comment:
I would reshape the code to put the `if` branch out. Something like this:
```c
    int f128_is_aligned = 1;
    if ((uintptr_t)f128_buf_to & 0xF) {
        f128_is_aligned = 0;
    }
    if( ldbl_is_aligned && f128_is_aligned ) {
        do {
            *(__float128*)f128_buf_to = *(long double*)ldbl_buf_from;
            ldbl_buf_from += from_extent;
            f128_buf_to += sizeof(__float128);
            count--;
        } while (count > 0);
    } else {
        do {
            if (((uintptr_t)ldbl_buf_from & alignment_mask) && f128_is_aligned) {
                *(__float128*)f128_buf_to = *(long double*)ldbl_buf_from;
            } else {
                __float128 f128;
                long double ldbl;
                memcpy(&ldbl, ldbl_buf_from, sizeof(ldbl));
                f128 = ldbl;
                memcpy(f128_buf_to, &f128, sizeof(f128));
            }

            ldbl_buf_from += from_extent;
            f128_buf_to += sizeof(__float128);
            count--;
        } while (count > 0);
    }
```

Reply:
Is that change for readability?  I'm assuming compilers are smart enough it wouldn't matter for performance.  I do like how the first section becomes compact, and then a reader can see that the second section is basically the same just handling bad alignment, but I don't like that it's duplicating code for the buffer/count updates at the bottom of each iteration.

Reply:
My first comment (taking the alignment_mask in account) was about correctness. If you make that change then the compiler cannot optimize the code because the condition is changing between iterations, so the generate code will have extra branches. I looked at the assembly on godbolt and the code looks more optimized with the code I pasted.

Reply:
I agree on the alignment check and pushed a change for that using a simple up-front check.  The downside of the single check is just for example if the from_extent was 12, then every other double could be an assignment and the other "every other double" could still be a memcpy.  But that's a lot of extra checking to make that happen, so I'm guessing the single up-front check which would turn the 12byte case into mecpy,memcpy,memcpy... is better.